### PR TITLE
[language] Publish generic resources

### PR DESCRIPTION
**Pull request by dariorussi**
This allows for publishing of generic resources.
It's not as clean of a solution as we would like but we are going to change things significantly with the binary format changes and the loader work.

### CODE REVIEW

File: language/functional_tests/tests/testsuite/payments/print_counterfeit_money.mvir [hunk 1]
```diff
@@ -48,6 +48,4 @@ main() {
     return;
 }
 
-// check: TYPE_ERROR
-// check: TYPE_ERROR
-// check: LINKER_ERROR
+// check: MISSING_DATA
```
Comment (dariorussi) [inline language/functional_tests/tests/testsuite/payments/print_counterfeit_money.mvir]: this should be the right error right? I mean I know it is because the test passes but it's also the expected one, is it?

File: language/bytecode-verifier/src/signature.rs [hunk 2]
```diff
@@ -199,25 +199,11 @@ impl<'a> SignatureChecker<'a> {
                                         self.module.struct_handle_at(struct_def.struct_handle);
                                     let type_actuals =
                                         &self.module.locals_signature_at(*type_actuals_idx).0;
-                                    let mut errs = check_generic_instance(
+                                    check_generic_instance(
                                         context,
                                         &struct_handle.type_formals,
                                         type_actuals,
-                                    );
```
Comment (dariorussi) [inline language/bytecode-verifier/src/signature.rs]: anything else we need to do here for the verifier?

File: language/vm/vm-runtime/src/runtime.rs [hunk 3]
```diff
@@ -216,6 +226,17 @@ impl<'alloc> VMRuntime<'alloc> {
             .resolve_function_ref(caller_module, idx, data_view)
     }
 
+    pub fn resolve_signature_token(
```
Comment (dariorussi) [inline language/vm/vm-runtime/src/runtime.rs]: this is a convenient function for now so I'd like to remove it another time

File: language/vm/vm-runtime/src/interpreter.rs [hunk 4]
```diff
@@ -914,10 +953,11 @@ impl<'txn> Interpreter<'txn> {
         current_frame: &Frame<'txn, FunctionRef<'txn>>,
     ) -> VMStatus {
         if err.is(StatusType::InvariantViolation) {
+            let state = self.get_internal_state(current_frame);
```
Comment (dariorussi) [inline language/vm/vm-runtime/src/interpreter.rs]: no reason for this except it makes it easier to `println!()` it if we need it for debugging reasons

File: language/vm/vm-runtime/src/identifier.rs [hunk 5]
```diff
@@ -5,12 +5,16 @@
...
     access_path::{AccessPath, Accesses},
     account_address::AccountAddress,
-    language_storage::{ResourceKey, StructTag},
+    language_storage::{ResourceKey, StructTag, TypeTag},
 };
 use vm::{access::ModuleAccess, file_format::StructDefinitionIndex};
 
 /// Get the StructTag for a StructDefinition defined in a published module.
-pub fn resource_storage_key(module: &impl ModuleAccess, idx: StructDefinitionIndex) -> StructTag {
+pub fn resource_storage_key(
```
Comment (dariorussi) [inline language/vm/vm-runtime/src/identifier.rs]: this is the change needed for the access path

File: language/vm/vm-runtime/src/interpreter.rs [hunk 6]
```diff
@@ -259,6 +260,18 @@ impl<'txn> Interpreter<'txn> {
                             )
                         })
                         .collect::<VMResult<Vec<_>>>()?;
+                    let type_context = TypeContext::new(current_frame.type_actuals().to_vec());
```
Comment (dariorussi) [inline language/vm/vm-runtime/src/interpreter.rs]: so setting up a call frame now computes both the instantiation for the types and the type tags.
That is going to change as soon as we do the loader work and the binary format changes

File: language/vm/vm-runtime/src/interpreter.rs [hunk 7]
```diff
@@ -811,8 +837,20 @@ impl<'txn> Interpreter<'txn> {
             StructDef,
         ) -> VMResult<AbstractMemorySize<GasCarrier>>,
     {
-        let ap = Self::make_access_path(module, idx, address);
-        let struct_def = runtime.resolve_struct_def(module, idx, context)?;
+        let module = frame.module();
```
Comment (dariorussi) [inline language/vm/vm-runtime/src/interpreter.rs]: this is where the substitution of the instantiation in the instructions (the signature in the global* bytecodes) is happening in order to generate a proper instantiation

**Comment by dariorussi** [comment]
I am seeing a problem with this as I expand test coverage so changes will be coming. 
But please take a look and see if the overall story is reasonable to you

File: language/vm/vm-runtime/src/interpreter.rs (hunk 7 above)
Comment (dariorussi) [reply language/vm/vm-runtime/src/interpreter.rs]: > this is where the substitution of the instantiation in the instructions (the signature in the global* bytecodes) is happening in order to generate a proper instantiation

actually it looks good and thanks to @vgao1996 to cover my insanity in writing the test... :)

File: language/functional_tests/tests/testsuite/generics/resource_safety/borrow_global.mvir [hunk 8]
```diff
@@ -1,18 +0,0 @@
-module M {
-    resource Foo<T> { x: T }
```
Comment (sblackshear) [inline language/functional_tests/tests/testsuite/generics/resource_safety/borrow_global.mvir]: Can we keep this test, but do `resource Foo<T>` -> `struct Foo<T>` (and similar for the tests below)?

File: language/bytecode-verifier/src/signature.rs (hunk 2 above)
Comment (sblackshear) [reply language/bytecode-verifier/src/signature.rs]: Don't know the verifier code well enough to say for sure, but I think it would be useful to test corner cases around "inferred resources":

```
module M {
struct MaybeResource<Type> { t: Type }

public yes1(v: LibraCoin.T) {
  move_to_sender<MaybeResource<LibraCoin.T>>(MaybeResource { t: v });
}

public yes2<R: resource>(r: R) {
  move_to_sender<MaybeResource<R>>(MaybeResource { t: r });
}

public no1(v: u64) {
  move_to_sender<MaybeResource<u64>>(MaybeResource { t: v });
}

public no2<A>(a: A) {
  move_to_sender<MaybeResource<A>>(MaybeResource { t: a });
}

public no3<C: copyable>(c: C) {
  move_to_sender<MaybeResource<C>>(MaybeResource { t: c });
}
}
```

File: language/functional_tests/tests/testsuite/payments/print_counterfeit_money.mvir (hunk 1 above)
Comment (sblackshear) [reply language/functional_tests/tests/testsuite/payments/print_counterfeit_money.mvir]: @vgao1996 should know for sure, but this looks right to me.

**Comment by dariorussi** [comment]
expanded the test, adding more

**Comment by dariorussi** [comment]
> I am seeing a problem with this as I expand test coverage so changes will be coming.
> But please take a look and see if the overall story is reasonable to you

actually it looks good and thanks to @vgao1996 to cover my insanity in writing the test... :)

File: language/vm/vm-runtime/src/interpreter.rs [hunk 9]
```diff
@@ -1066,6 +1106,7 @@ struct Frame<'txn, F: 'txn> {
     locals: Locals,
     function: F,
     type_actual_tags: Vec<TypeTag>,
+    type_actuals: Vec<Type>,
```
[the remark this answers is not in the document]
Comment (vgao1996) [reply language/vm/vm-runtime/src/interpreter.rs]: I think this and many other scattered changes can be replaced by
```
fn derive_struct_layout_from_struct_tag(
    module: &LoadedModule,
    tag: &StructTag,
) -> VMResult<StructDef> {
    if tag.address != *module.address() || tag.module.as_ident_str() != module.name() {
        return Err(VMStatus::new(StatusCode::INTERNAL_TYPE_ERROR));
    }
    let idx = module.get_struct_def_index(tag.name.as_ident_str())?;
    let mut arg_layouts = tag
        .type_params
        .iter()
        .map(|arg_tag| derive_layout_from_tag(module, arg_tag))
        .collect::<VMResult<Vec<_>>>()?;
    let def = module.struct_def_at(*idx);
    match def.field_information {
        StructFieldInformation::Native => {
            // Currently only 1 native type is allowed: Vector.T
            if tag.module.as_ident_str() != IdentStr::new("Vector").unwrap()
                || tag.name.as_ident_str() != IdentStr::new("T").unwrap()
            {
                return Err(VMStatus::new(StatusCode::INTERNAL_TYPE_ERROR));
            }
            // Vector.T must take exactly 1 type argument.
            if arg_layouts.len() != 1 {
                return Err(VMStatus::new(StatusCode::INTERNAL_TYPE_ERROR));
            }
            Ok(StructDef::Native(NativeStructType::new_vec(
                arg_layouts.pop().unwrap(),
            )))
        }
        StructFieldInformation::Declared {
            field_count,
            fields,
        } => {
            let field_layouts = module
                .field_def_range(field_count, fields)
                .iter()
                .map(|fd| {
                    let ty = &module.type_signature_at(fd.signature).0;
                    let tag = derive_type_tag(module, &tag.type_params, ty)?;
                    let layout = derive_layout_from_tag(module, &tag)?;
                    Ok(layout)
                })
                .collect::<VMResult<Vec<_>>>()?;
            Ok(StructDef::new(field_layouts))
        }
    }
}

fn derive_layout_from_tag(module: &LoadedModule, tag: &TypeTag) -> VMResult<Type> {
    let res = match tag {
        TypeTag::Bool => Type::Bool,
        TypeTag::U8 => Type::U8,
        TypeTag::U64 => Type::U64,
        TypeTag::U128 => Type::U128,
        TypeTag::Address => Type::Address,
        TypeTag::ByteArray => Type::ByteArray,
        TypeTag::Struct(tag) => Type::Struct(derive_struct_layout_from_struct_tag(module, tag)?),
    };

    Ok(res)
}
```
(Code compiles, but *untested*.)

The idea is that once you have the struct tag of the resource being published, you can recursively look up its layout. I personally prefer this solution as it is more self-contained and avoids have duplicate logic for substitution.

**Comment by dariorussi** [comment]
just updated the test to make sure the feature has basic functionality.
We publish several generic resources and we go through a lifetime of creation, usage, destruction.
I will look at the feedback now

**Comment by dariorussi** [comment]
added back some of the tests suggested by @sblackshear and that felt what was needed short of a more structure testing strategy which will have to come at some point. Not this PR though

File: language/bytecode-verifier/src/signature.rs (hunk 2 above)
Comment (dariorussi) [reply language/bytecode-verifier/src/signature.rs]: it turns out those are all invalid examples because we can only publish declared resources and not inferred resources

Comment (sblackshear) [reply language/bytecode-verifier/src/signature.rs]: Ah, good catch! This seems like a very sensible policy to me, but it would interesting to explore whether it would be sound to allow examples like `yes1` and `yes2` (just for our own understanding, not because I think we need this feature).

File: language/vm/vm-runtime/src/interpreter.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (dariorussi) [reply language/vm/vm-runtime/src/interpreter.rs]: this is a good suggestion however I'd like to make this (or similar) going into the coming rewrite.
Basically I'd like to commit this as is and do improvements next.
@vgao1996 let me know your thoughts, meanwhile, I keep thinking how to plug this is
